src/utils/async.ts

```diff
@@ -48,3 +48,109 @@ export async function asyncIterableToArray<T>(
   const iter = source[Symbol.asyncIterator]();
   return asyncIteratorToArray(iter);
 }
+
+/**
+ * Create a new AsyncIterable providing only the values that satisfy the predicate function.
+ * The predicate function is also an optional Type Guard function if types T and K are different.
+ *
+ * Example:
+ * ```
+ * const i: AsyncIterable<Either<E, A>> = {} as AsyncIterable<Either<E, A>>;
+ * const f: AsyncIterable<Right<E, A>> = filterAsyncIterable<Either<E, A>, Right<E, A>>(i, isRight);
+ * ```
+ * @param iterable Original AsyncIterable
+ * @param predicate Predicate function
+ */
+export const filterAsyncIterable = <T, K = T>(
+  iterable: AsyncIterable<T | K>,
+  predicate: (value: T | K) => value is K
+): AsyncIterable<K> => ({
+  async *[Symbol.asyncIterator](): AsyncIterator<K> {
+    // tslint:disable-next-line: await-promise
+    for await (const value of iterable) {
+      if (predicate(value)) {
+        yield value;
+      }
+    }
+  }
+});
+
+/**
+ * Create a new AsyncIterator providing only the values that satisfy the predicate function.
+ * The predicate function is also an optional Type Guard function if types T and K are different.
+ *
+ * Example:
+ * ```
+ * const i: AsyncIterator<Either<E, A>> = {} as AsyncIterator<Either<E, A>>;
+ * const newI: AsyncIterator<Right<E, A>> = filterAsyncIterator<Either<E, A>, Right<E, A>>(iterable, isRight);
+ * ```
+ * @param iter Original AsyncIterator
+ * @param predicate Predicate function
+ */
+export function filterAsyncIterator<T, K = T>(
+  iter: AsyncIterator<T | K>,
+  predicate: (value: T | K) => value is K
+): AsyncIterator<K> {
+  const iterable = {
+    [Symbol.asyncIterator]: () => iter
+  };
+  return filterAsyncIterable(iterable, predicate)[Symbol.asyncIterator]();
+}
+
+/**
+ * Create a new AsyncIterable which provide one by one the values ​​contained into the input AsyncIterable
+ *
+ * @param iterable Original AsyncIterable
+ */
+export const flattenAsyncIterable = <T>(
+  iterable: AsyncIterable<ReadonlyArray<T>>
+): AsyncIterable<T> => ({
+  async *[Symbol.asyncIterator](): AsyncIterator<T> {
+    // tslint:disable-next-line: await-promise
+    for await (const value of iterable) {
+      for (const item of value) {
+        yield item;
+      }
+    }
+  }
+});
+
+/**
+ * Create a new AsyncIterator which provide one by one the values ​​contained into the input AsyncIterator
+ *
+ * @param iter Original AsyncIterator
+ */
+export function flattenAsyncIterator<T>(
+  iter: AsyncIterator<ReadonlyArray<T>>
+): AsyncIterator<T> {
+  const iterable = {
+    [Symbol.asyncIterator]: () => iter
+  };
+  return flattenAsyncIterable(iterable)[Symbol.asyncIterator]();
+}
+
+export function reduceAsyncIterable<A, B>(
+  iterable: AsyncIterable<ReadonlyArray<A>>,
```

Comment:
This should take an AsyncIterable<A> (we can flatten AsyncIterable<ReadonlyArray<A>> to get an AsyncIterable<A>)

Reply:
If we flatten `AsyncIterable<ReadonlyArray>` the output will be different. This function reduces the single `ReadonlyArray` to one value of the AsyncIterator.
I think that this test explain what I mean [link](https://github.com/pagopa/io-functions-commons/pull/82/files#diff-bc0a6d43f117d3c7951c79b835270311R128-R152)

Reply:
ah I see, the name is misleading then, and you don't need a specific method to do that, you can map the iterator with a reduce:

```ts
mapAsyncIterator(iter, _ => _.reduce(f, zero))
```

Reply:
Nice :), so I'll remove the code.